### ISSUE TITLE
Build fails with gcc in strict c11 mode due to reliance on typeof compiler extension

### ISSUE THREAD

**Issue by wisp3rwind** [axboe/liburing#181]
When building a project that includes `liburing` using gcc 10.2.0 with the `--std=c11` option, the compiler fails with several errors similar to the following:
```
In file included from /usr/include/liburing/barrier.h:56,
                 from /usr/include/liburing.h:16,
                 from ../src/main.c:13:
/usr/include/liburing.h: In function ‘io_uring_cq_advance’:
/usr/include/liburing.h:150:3: error: unknown type name ‘typeof’
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: error: expected ‘)’ before ‘cq’
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: error: expected ‘)’ before ‘*’ token
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: error: expected ‘)’ before ‘(’ token
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: note: to match this ‘(’
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: error: ‘__atomic_store_ptr’ undeclared (first use in this function); did you mean ‘__atomic_store_n’?
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/liburing.h:150:3: note: each undeclared identifier is reported only once for each function it appears in
/usr/include/liburing.h:150:3: error: argument 1 of ‘__atomic_store’ must be a non-void pointer type
  150 |   io_uring_smp_store_release(cq->khead, *cq->khead + nr);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

The reason is the usage of the `typeof` extension in `liburing/barrier.h`, which gcc disables when asking for strict standard compliance. However, simply replacing `typeof` with `__typeof__` (see [Alternate names](https://gcc.gnu.org/onlinedocs/gcc/Typeof.html)) lets the build succeed. Thus, `liburing` appears to be standard-compliant otherwise. Hence, I propose to upstream this substitution. According to [the docs](https://clang.llvm.org/docs/UsersManual.html#differences-between-various-standard-modes), clang supports the alternate name, too.

**Comment by axboe** [comment]
I'll be honest with you, I care a lot more about building with that people have than some supposed standards compliance. I'm fine with making it fully c11 (or whatever) compliant if it doesn't break anything else, obviously, but it not a goal in and of itself for me.

Does __typeof__ work with gcc-4.9? Older clang?

Also, the build will still fail with --std=c11 even with the typeof -> __typeof__ change.

**Comment by wisp3rwind** [comment]
> I'll be honest with you, I care a lot more about building with that people have than some supposed standards compliance. I'm fine with making it fully c11 (or whatever) compliant if it doesn't break anything else, obviously, but it not a goal in and of itself for me.

I'm on your side here, this not a goal in and of itself for me, either. I just happened to have set `c_std=c11` in a meson config copied from somewhere, and it took quite some time to figure out what was going on. My initial thought was that, if this is indeed the only thing preventing `liburing` from building with strict c11 enforced, there'd be little reason not to make the change.

> Does `__typeof__` work with gcc-4.9? Older clang?
> 
> Also, the build will still fail with --std=c11 even with the `typeof` -> `__typeof__` change.

I doubt I'll really invest the time to check compiler support in detail, so I'll close the issue again. Still, someone might find it useful since it is now (hopefully) easier to find some information about the problem using google/github search.

**Comment by axboe** [comment]
> Still, someone might find it useful since it is now (hopefully) easier to find some information about the problem using google/github search.

Agree, that's always useful!
